### PR TITLE
Fix stale relay list: replace UserDefaults hex lookup with ndb query

### LINKED ISSUE #3537: relay list disco
bug report:

1. Uh, guise...my relays are spazzing out. I can't seem to remove the ones showing, and we'll, just see video
2. relays in relay management view are flashing (see videos)

[https://v.nostr.build/69Dqh1sKtZcPUtQ4.mp4](<https://v.nostr.build/69Dqh1sKtZcPUtQ4.mp4>)<br>[https://v.nostr.build/ZSUZVEqpOIW2szT4.mp4](<https://v.nostr.build/ZSUZVEqpOIW2szT4.mp4>)

[https://damus.io/nevent1qqs960j8cs7g9vr7ca8wm3na3qyx2x9hn3xv7wgp5cgydk8w33chufc89cat2](<https://damus.io/nevent1qqs960j8cs7g9vr7ca8wm3na3qyx2x9hn3xv7wgp5cgydk8w33chufc89cat2>)

1.16 (1264) [https://github.com/damus-io/damus/commit/b105dadd1482140f5603cc8891653d9191600381](<https://github.com/damus-io/damus/commit/b105dadd1482140f5603cc8891653d9191600381>)<br>iOS 26.2

reported by npub162zpxufpw8pnuytaf0gfxzkqtvk9rvcwkvppa7x57y3n7qkfpg4shatdhy

### PR DESCRIPTION
**Pull request by alltheseas**
## Summary

Replace the fragile `latestRelayListEventIdHex` UserDefaults lookup in `getLatestNIP65RelayListEvent()` with a direct nostrdb query using `ndb.query(filters:maxResults:)` on the AUTHOR_KINDS index.

The old hex-based lookup could go stale across sessions or when the listener's race condition prevented proper updates. When this happened, relay add/remove operations silently fell back to bootstrap or year-old relay lists — making it impossible to disconnect relays (#3537).

**Key changes:**
- **ndb query instead of hex lookup** — `getLatestNIP65RelayListEvent()` now queries nostrdb for the latest kind:10002 event by author, eliminating the stale-hex failure mode
- **In-memory cache** — `lastSetRelayList` bridges the nostrdb async write gap: `set()` populates it, `getUserCurrentRelayList()` checks it first, `load()` clears it
- **Removed `latestRelayListEventIdHex`** — from Delegate protocol, DamusState, SaveKeysView, UserSettingsStore, and all test mocks

Closes #3537

## Commits (review commit-by-commit)

1. `60d73879` — Fix stale relay list: replace UserDefaults hex lookup with ndb query (19 impl, 198 test)

Total: 19 impl, 198 test

## Checklist

- [x] I have read (or I am familiar with) the [Contribution Guidelines](../docs/CONTRIBUTING.md)
- [x] I have tested the changes in this PR
- [x] I have profiled the changes to ensure there are no performance regressions, or I do not need to profile the changes.
    - Not needed: changes are control-flow only (query method swap, in-memory cache) — no new allocations or hot-path logic.
- [x] I have opened or referred to an existing github issue related to this change.
    - #3537
- [x] My PR is either small, or I have split it into smaller logical commits that are easier to review
- [x] I have added the signoff line to all my commits. See [Signing off your work](../docs/CONTRIBUTING.md#sign-your-work---the-developers-certificate-of-origin)
- [x] I have added appropriate changelog entries for the changes in this PR. See [Adding changelog entries](../docs/CONTRIBUTING.md#add-changelog-changed-changelog-fixed-etc)
    - `Changelog-Fixed: Fix stale relay list causing inability to disconnect relays`
- [x] I have added appropriate `Closes:` or `Fixes:` tags in the commit messages wherever applicable

## Test report

**Device:** iPhone 17 Pro Simulator

**iOS:** 26.2

**Damus:** This branch

**Tests run:** `NostrNetworkManagerTests` (8 tests), `EntityPreloaderTests` (9 tests), `SubscriptionManagerNegentropyTests` (7 tests)

**Results:** All 24 tests pass.

### Regression tests (5 new)

| Test | Scenario |
|------|----------|
| `testRemoveRelayDoesNotFallBackToBootstrapList` | Remove relay B from [A,B,C] → must yield [A,C], not bootstrap |
| `testCacheBridgesAsyncWriteGap` | `set()` then immediate `getUserCurrentRelayList()` → cache serves new list |
| `testRapidSequentialRemovesDoNotReintroduceRelays` | Remove B then C from [A,B,C] → result is [A], B doesn't reappear |
| `testLoadClearsCacheAndReadsFromNdb` | `load()` clears cache so ndb becomes source of truth |
| `testNdbQueryFindsRelayListWithoutStoredHex` | ndb query finds relay list without any stored hex ID |

### Before/After proof (AGENTS.md Rule 17)

All 5 regression tests were verified by reverting the fix and confirming failure:

| Test | Before (reverted) | After (fixed) |
|------|-------------------|---------------|
| `testRemoveRelayDoesNotFallBackToBootstrapList` | FAIL — `remove()` throws, no relay list found | PASS |
| `testCacheBridgesAsyncWriteGap` | FAIL — `XCTAssertNotNil failed` | PASS |
| `testRapidSequentialRemovesDoNotReintroduceRelays` | FAIL — `remove()` throws | PASS |
| `testNdbQueryFindsRelayListWithoutStoredHex` | FAIL — `XCTAssertNotNil failed` | PASS |
| `testLoadClearsCacheAndReadsFromNdb` | FAIL — `set()` throws | PASS |

## Summary by CodeRabbit

* **Bug Fixes**
  * Added an in-memory cache and cache-first, query-based relay-list retrieval to prevent stale data, ensure immediate updates, and avoid reintroducing removed relays; cache is cleared on load.

* **Tests**
  * Added comprehensive relay-list regression tests covering rapid updates, cache bridging, removal edge cases, and query scenarios.

* **Chores**
  * Removed legacy relay-list persistence hooks and related test-only helper state.

### CODE REVIEW

**Comment by alltheseas** [comment]
> replaceable events nostrdb - query replaceable

> whiles its doing query, make sure its select latest one

> nostrdb will gave the latest on query, but if you are subscribing to remote server, first result is not necessarily latest

@jb55 asked to check

File: damus/Core/Networking/NostrNetworkManager/UserRelayListManager.swift [hunk 1]
```diff
@@ -80,17 +86,18 @@ extension NostrNetworkManager {
...
         ///
         /// - Returns: The latest NIP-65 relay list NdbNote
         private func getLatestNIP65RelayListEvent() -> NdbNote? {
-            guard let latestRelayListEventId = delegate.latestRelayListEventIdHex else { return nil }
-            guard let latestRelayListEventId = NoteId(hex: latestRelayListEventId) else { return nil }
-            return try? delegate.ndb.lookup_note_and_copy(latestRelayListEventId)
+            let filter = NostrFilter(kinds: [.relay_list], limit: 1, authors: [delegate.keypair.pubkey])
+            guard let ndbFilter = try? NdbFilter(from: filter) else { return nil }
+            guard let noteKey = try? delegate.ndb.query(filters: [ndbFilter], maxResults: 1).first else { return nil }
+            return try? delegate.ndb.lookup_note_by_key_and_copy(noteKey)
```
[the remark this answers is not in the document]
Comment (danieldaquino) [reply damus/Core/Networking/NostrNetworkManager/UserRelayListManager.swift:101]: This is fine, ndb should be fast enough to handle these. I think we should loosen this review guideline. This kind of query on the main thread is not _universally_ bad. Only in some cases where it might run longer or if it's a performance critical path.
